report/04_implementation/01_flutter_application.md

```diff
@@ -6,23 +6,150 @@ web application) with a single code base.
 Because of this, the development does not take place platform-specifically,
 but in one programming language and environment.
 
-The programming language used with *Flutter* is *dart*. 
+The programming language used with *Flutter* is *Dart*. 
 *Dart* is an object oriented programming language that is very similar to C#
 or Java. 
 
-// TODO: Mehr Informationen zu DART / Flutter
-
+Its advantage over Java, C# etc. is a modern implementation where common 
+issues have been fixed, that older languages weren't aware of while being 
+developed. Dart for example implements null sound safety null checker where 
+a developer has to ensure that all variables contain non null values. 
+Otherwise the program will not compile.  
+This is a big advantage over Java since developers don't have to handle 
+*NullPointerException*s at runtime and don't risk forgetting them.
 
 All in all we tried to set up a reusable folder and file structure with
-seperation of concerns between the frontend / user interface and the business
+separation of concerns between the frontend / user interface and the business
 logic.
 
-### Config
-
-### Database
+### Service Provider
+
+The key part of the app is provided by a service provider. This service 
+provider ensures that all key elements like the config manager, the database 
+manager and notifications have been initialized successfully before enabling 
+the ui and allowing the user to interact with the app.  
+During runtime of the app the service provider can be accessed from everywhere 
+within the app's source code with help from the 
+[provider](https://pub.dev/packages/provider) package. This has the advantage 
+of utilizing the Singleton design pattern principle for all key parts of the 
+app.  
+While the provider package ensures the existence of only one service provider 
+instance, a critical issue arises when initializing all key parts. This is 
+mostly because of the asynchronous execution of code parts which have to be 
+synchronized. Solving this issue needed some sort of *Mutex* to be 
+implemented.  
+Unfortunately, *Dart* is kind of restrictive when it comes to using 
+rudimentary programming principles like *Mutexes*. Therefore, the `init` 
+method of the service provider takes a so called `Closure` which can be 
+described as nameless function. This function is invoked after all 
+initialization tasks have been completed.  
+The code example below explains this process in a more detailed manner.
+
+```dart
```

Comment:
I think since the code is hear meant to be read by not programmers it would be kind to add more comments explaining more what is going on in non programmer understandable text :)

Reply:
Imo more comments contradict the fact of adding code examples to this report. Since the basic functionality is already explained as plain text this code example is only an enhancement for programmers who are interested in further understanding how this can be implemented in Dart.

In short, non programmers will most likely skip the code example ^^

Reply:
sure if there are no comments that help them to understand it ;)
Also IMHO your last comment contradicts the line above the code example:
```
The code example below explains this process in a more detailed manner.
```
Since you only show the code and don't explain it in more detail.